zowe-rest-api-sample-spring/docs/api-client-authentication.md

```diff
@@ -0,0 +1,201 @@
+
+This section focuses on the way how services in the Zowe API ecosystem are expected to accept and use these tokens so the API clients have unified experience.
+
+### Token-based Login Flow and Request/Response Format
+
+1. The API client obtains JWT token by using POST method on `/auth/login` endpoint of API service that requires valid user ID and password
+
+2. The API service calls an *authentication provider* that returns a JWT token that contains the user ID claim in the HTTP cookie named `apimlAuthenticationToken` with attributes `HttpOnly` and `Secure`.
+
+3. The API client remembers the JWT token or cookie and sends it with every request as the cookie with name `apimlAuthenticationToken`.
+
+#### Obtaining the token
+
+- The full URL is the base URL of the API service plus `/auth/login`. If the application has the base URL with `/api/v1` then the full URL can be: `https://hostname:port/api/v1/auth/login`.
+
+- The credentials are provide in JSON request:
+
+    ```json
+    {
+        "username": "...",
+        "password": "..."
+    }
+    ```
+
+- Successful login returns `204` and empty body with the token in the `apimlAuthenticationToken` cookie
+
+- Failed authentication returns `401` without `WWW-Authenticate`
+
+**Example**:
+
+```bash
+curl -v -c - -X POST "https://localhost:10080/api/v1/auth/login" -d "{ \"username\": \"zowe\", \"password\": \"zowe\"}"
+```
+
+```http
+POST /api/v1/auth/login HTTP/1.1
+Accept: application/json, */*
+Content-Length: 40
+Content-Type: application/json
+
+{
+    "username": "zowe",
+    "password": "zowe"
+}
+
+HTTP/1.1 204
+Set-Cookie: apimlAuthenticationToken=eyJhbGciOiJSUzI1NiJ9...; Path=/; Secure; HttpOnly
+```
+
+#### Authenticated request
+
+The API client just pass the JWT token as a Cookie header with name `apimlAuthenticationToken`:
+
+```http
+GET /api/v1/greeting HTTP/1.1
+Cookie: apimlAuthenticationToken=eyJhbGciOiJSUzI1NiJ9...
+
+HTTP/1.1 200
+...
+```
+
+### Token format
+
+The JWT must contain unencrypted claims `sub`, `iat`, `exp`, `iss`, and `jti` in the meaning defined by <https://tools.ietf.org/html/rfc7519#section-4.1>. Specifically, the `sub` is the z/OS user ID and `iss` is the name of the service that issued the JWT token.
+
+The JWT must use RS256 signature algorithm and the secret used to sign the JWT is an asymmetric key generated during installation.
+
+**Example**:
+
+```json
+{
+  "sub": "zowe",
+  "iat": 1575034758,
+  "exp": 1575121158,
+  "iss": "Zowe Sample API Service",
+  "jti": "ac2eb63e-caa6-4ccf-a527-95cb61ad1646"
+}
+```
+
+### Validating tokens
+
+API client does not need to validate the tokens, the API services must do it themselves. If the API client receives the token from another source or needs to check details in it (like user ID, expiration) then it can use `/auth/query` endpoint
```

Comment:
Does the query endpoint do more than check the details?  For example, does query validate it?

I was wondering if you can simply decode the token to check some details but you need query to validate it.

Reply:
/query always validates and decodes.

I will fix the sentence

Reply:
Thanks for the help